04_Scheduling/03_Tasks_And_Threads.md

```diff
@@ -0,0 +1,132 @@
+
+### Creating a thread
+
+When creating a thread we must first of all create it's execution frame (yeah even if not executed yet, it will need one).
+
+### Execution frame
+
+One of the way to implement stack switching is replacing the current iret frame with a the one contained in the next thread. To do that a task need when created to build an ad-hoc frame that will not cause any crash. What it should contain? Well this is basically a copy of the stack frame during the handling of an interrupt routine. 
+
+### What a task should contain ?
+
+So what are the information that we should store in a task? That depends on the design choices, because if we are implementing a multitask/single thread os it will need more information, but if we are going to implement a multi-task/multi-threaded OS some of the information will be stored in the Thread structure. So let's start with the minimum set of information that we need to save into a task: 
+
+* First of all a name, all tasks have a task name, most of the time is the excuatble name, sometime something different
+* Every task will have their own task id, this value usually is used by the scheduler to pick up the next task to execute or and in all the scenarios where an access to the task is needed (for example interprocess comunication)
+* A reference to it's own virtual addressing space, that is a pointer to a PML4 table (this only if we are going to implement an os where every task has it's own address space)
+
+### Context Switching
+
+* Usually context switch is done during an IRQ (when it depends on design decision. 
+
+One way to achieve context switching, is to clone the current stack frame for the IRQ being served and update some of it's values:
+
+* rip - this will contain the pointer to the thread function
+* rflgas - will containg 0x202
+* rsp - is the stack pointer, since the pointer is growing backward 
+* // to add other registers to initialize
+
+The rip will contain the function that is associated with the thread, sp contains the newly allocated stack for it. 
+
+### Exiting the thread
+
+After the thread finish its execution, we need a way to make it exit gracefully (otherwise it start to run into garbage most likely...) so there are two possible scenarios:
+
+* The programmer has called a thread_exit function so in this case we are fine
+* The programmer didn't called the function and the thread has terminated finished the execution of the function, at this point if it will not be stopped it will run into garbage. 
+
+To achieve that we need to have an "exit" function to be executed after the function in the thread terminates it's execution. But how to do it? We need to check on the X86_64 Abi calling convention (or what architecture applies to you). As usual there are multiple ways to achieve that, the easiest one is to create a wrapper function that takes two parameters
```

Comment:
I would split this into two separate steps:
- first you write the exit function, and supporting code. Show how it wold work with a thread calling it manually.
- Now you can introduce the idea of the wrapper function, and have it automatically exit when a thread's entry function returns.